File: app/schemas/tipsters.py
"""JSON shapes for Phase 6 tipsters / booking codes."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class TipsterCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128, examples=["Lagos Tips"])
    handle: str | None = Field(default=None, examples=["@lagostips"])
    platform: str | None = Field(
        default=None, examples=["instagram"], description="instagram|telegram|twitter|other"
    )
    notes: str | None = None


class TipsterOut(BaseModel):
    id: int
    name: str
    handle: str | None = None
    platform: str | None = None
    notes: str | None = None
    created_at: datetime | None = None
    codes_total: int | None = None
    codes_pending: int | None = None
    codes_settled: int | None = None


class CodeCreate(BaseModel):
    tipster_id: int
    code_text: str = Field(min_length=1, max_length=256, examples=["ABC123XYZ"])
    bookmaker: str = Field(default="sportybet", examples=["sportybet", "bet9ja"])
    stake_ngn: Decimal | None = Field(default=None, gt=0)
    odds_price: Decimal | None = Field(default=None, gt=1)
    source: str = Field(default="manual", examples=["manual", "instagram", "telegram"])
    notes: str | None = Field(
        default=None,
        description="Optional slip text — we extract market hints (1X, O2.5, BTTS…)",
    )
    markets_summary: str | None = None


class CodeOut(BaseModel):
    id: int
    tipster_id: int
    tipster_name: str
    code_text: str
    bookmaker: str
    markets_summary: str | None = None
    stake_ngn: Decimal | None = None
    odds_price: Decimal | None = None
    source: str
    notes: str | None = None
    result: str
    created_at: datetime | None = None
    settled_at: datetime | None = None


class CodeSettleRequest(BaseModel):
    result: str = Field(examples=["won"], description="won | lost | void | pending")


class CodeCreateResponse(BaseModel):
    status: str
    code: CodeOut | None = None
    message: str


class LeaderboardResponse(BaseModel):
    count: int
    min_settled: int
    leaderboard: list[dict]
    message: str

File: mobile/src/api/tipsters.ts
import { getJson, postJson } from './client';

export type Tipster = {
  id: number;
  name: string;
  handle?: string | null;
  platform?: string | null;
  notes?: string | null;
  codes_total?: number | null;
  codes_pending?: number | null;
  codes_settled?: number | null;
};

export type BookingCode = {
  id: number;
  tipster_id: number;
  tipster_name: string;
  code_text: string;
  bookmaker: string;
  markets_summary?: string | null;
  stake_ngn?: number | string | null;
  odds_price?: number | string | null;
  source?: string;
  notes?: string | null;
  result: string;
  created_at?: string | null;
  settled_at?: string | null;
};

export type LeaderboardRow = {
  tipster_id?: number;
  name?: string;
  handle?: string | null;
  settled?: number;
  won?: number;
  lost?: number;
  hit_rate_pct?: number | null;
  roi_pct?: number | null;
  [key: string]: unknown;
};

export function fetchTipsters(limit = 50) {
  return getJson<Tipster[]>(`/tipsters?limit=${limit}`);
}

export function createTipster(body: {
  name: string;
  handle?: string | null;
  platform?: string | null;
  notes?: string | null;
}) {
  return postJson<Tipster>('/tipsters', body);
}

export function fetchCodes(opts?: { tipster_id?: number; result?: string; limit?: number }) {
  const q = new URLSearchParams();
  if (opts?.tipster_id != null) q.set('tipster_id', String(opts.tipster_id));
  if (opts?.result) q.set('result', opts.result);
  q.set('limit', String(opts?.limit ?? 50));
  return getJson<BookingCode[]>(`/tipsters/codes?${q}`);
}

export function logBookingCode(body: {
  tipster_id: number;
  code_text: string;
  bookmaker?: string;
  stake_ngn?: number | null;
  odds_price?: number | null;
  notes?: string | null;
  source?: string;
}) {
  return postJson<{ status: string; code: BookingCode | null; message: string }>(
    '/tipsters/codes',
    {
      tipster_id: body.tipster_id,
      code_text: body.code_text,
      bookmaker: body.bookmaker || 'sportybet',
      stake_ngn: body.stake_ngn ?? null,
      odds_price: body.odds_price ?? null,
      notes: body.notes || null,
      source: body.source || 'manual',
    }
  );
}

export function settleBookingCode(codeId: number, result: string) {
  return postJson<BookingCode>(`/tipsters/codes/${codeId}/settle`, { result });
}

export function fetchTipsterLeaderboard(minSettled = 1) {
  return getJson<{
    count: number;
    min_settled: number;
    leaderboard: LeaderboardRow[];
    message: string;
  }>(`/tipsters/leaderboard?min_settled=${minSettled}`);
}
